### PR TITLE
Improve PPS's fault tolerance

### LINKED ISSUE #762: Robust pipeline error detection
Currently, when something goes wrong in a pod, we rely on the pod itself to report that it has failed, at which point we mark the job as failed as well.  However, if a pod for whatever reason fails to contact PPS, we don't have any mechanism in place that would proactively check on the pod, so the job will just forever be stuck in the "RUNNING" state, even though it's not actually making progress.

### PR DESCRIPTION
**Pull request by derekchiang**
This PR revamps PPS's job execution model.  The end results are:

* Jobs can now survive crashes (i.e. failures that are not explicitly reported to PPS), including pod crashes and node crashes.
* Jobs are more resilient to transient failures: when the user container fails, PPS reschedules the container for up to 3 times before giving up.
* We have a good foundation for future work such as 1) having each pod work on multiple chunks of data, 2) straggler detection and mitigation (#442), 3) work recovery from failed jobs (#606) etc.

This PR seems to fix #762 and #905, and partially addresses #990.  I might be missing a couple more.

### CODE REVIEW

File: src/server/pps/server/api_server.go [hunk 1]
```diff
@@ -34,6 +35,12 @@ import (
 	kube_labels "k8s.io/kubernetes/pkg/labels"
 )
 
+const (
+	// MaxPodsPerChunk is the maximum number of pods we can schedule for each
+	// chunk in case of failures.
+	MaxPodsPerChunk = 3
```
Comment (jdoliner) [inline src/server/pps/server/api_server.go]: This is basically the maximum number of retries for a chunk right?
If so it may make sense to have that be a configurable part of the Job/Pipeline Spec.
This isn't strictly necessary for merging this code though.

File: src/server/pps/server/api_server.go [hunk 2]
```diff
@@ -1455,6 +1569,133 @@ func (a *apiServer) runPipeline(ctx context.Context, pipelineInfo *ppsclient.Pip
 	}
 }
 
+func (a *apiServer) jobManager(ctx context.Context, job *ppsclient.Job) error {
```
Comment (jdoliner) [inline src/server/pps/server/api_server.go]: This could use a more descriptive name and maybe a comment about what this function does.

File: src/server/pachyderm_test.go [hunk 3]
```diff
@@ -486,6 +485,116 @@ func TestPipeline(t *testing.T) {
...
+
+	c := getPachClient(t)
+
+	// create repos
+	dataRepo := uniqueString("TestPipelineThatCrashes_data")
+	require.NoError(t, c.CreateRepo(dataRepo))
+	// create pipeline
+	pipelineName := uniqueString("pipeline")
+	// this pipeline sleeps.
+	// then we are gonna manually kill a pod and see if the job completes successfully
```
Comment (derekchiang) [inline src/server/pachyderm_test.go]: oh yeah i was literally killing the pods manually.  should probably figure out a way to kill the pods programmatically.

File: src/server/pkg/lease/lease.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
+package lease
+
+import "time"
+
+// Manager manages resources via leases
+type Manager interface {
```
Comment (derekchiang) [inline src/server/pkg/lease/lease.go]: Yeah it was called `LeaseManager` then golint complains that `lease.LeaseManager` stutters.

Just tried `make lint` with `Leaser` and there was no complaint.  GG.

File: src/server/pps/pps.proto [hunk 5]
```diff
@@ -11,22 +11,42 @@ option go_package = "pps";
 
 message StartJobRequest {
```
Comment (derekchiang) [inline src/server/pps/pps.proto]: Makes sense.

File: src/server/pps/server/api_server.go (hunk 1 above)
Comment (derekchiang) [inline src/server/pps/server/api_server.go]: Agreed. We can do that in a separate PR.

File: src/server/pachyderm_test.go (hunk 3 above)
Comment (jdoliner) [inline src/server/pachyderm_test.go]: Yeah, otherwise we're going to need to setup CI to send you a slack message to go in and do your thing. I think there's examples in the tests of killing pachd pods that you can copy paste.

**Comment by derekchiang** [comment]
@jdoliner all comments addressed.

**Comment by jdoliner** [comment]
Looks like the one about sleeping in tests might have slipped through the cracks.

**Comment by jdoliner** [comment]
nine-and-watch
